core/domain/classifier_services.py

```diff
@@ -0,0 +1,534 @@
+        word instances assigned to each label for a given word id (aggregated
+        across all docs).
+    _c_l - a list of counts, where each count is the number of times
+        a label was assigned (aggregated over all word instances in all docs).
+
+    There are two internal learning rates, _DEFAULT_ALPHA and _DEFAULT_BETA.
+    These are initialized to Wikipedia's recommendations. Do not change these
+    unless you know what you're doing.
+
+    It is possible for a word instance in a doc to not have an explicit label
+    assigned to it. This is characterized by assigning _DEFAULT_LABEL to the
+    word instance.
+    """
+
+    _DEFAULT_ALPHA = 0.1
+    _DEFAULT_BETA = 0.001
+
+    _DEFAULT_TRAINING_ITERATIONS = 25
+    _DEFAULT_PREDICTION_ITERATIONS = 5
+
+    _DEFAULT_PREDICTION_THRESHOLD = 0.5
+
+    _DEFAULT_LABEL = '_default'
+
+    def __init__(self):
+        """
+        Initializes constants for the classifier. Setting a seed ensures
+        that results are deterministic. There is nothing special about the
+        value 4. These should not be changed unless you know what you're doing.
+        """
+        numpy.random.seed(seed=4)
+
+        self._alpha = self._DEFAULT_ALPHA
+        self._beta = self._DEFAULT_BETA
+
+        self._training_iterations = self._DEFAULT_TRAINING_ITERATIONS
+        self._prediction_iterations = self._DEFAULT_PREDICTION_ITERATIONS
+
+        self._prediction_threshold = self._DEFAULT_PREDICTION_THRESHOLD
+
+    def _get_word_id(self, word):
+        """
+        Returns a word's id if it exists, otherwise assigns
+        a new id to the word and returns it.
+        """
+        if word not in self._word_to_id:
+            self._word_to_id[word] = self._num_words
+            self._num_words += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """
+        Returns a label's id if it exists, otherwise assigns
+        a new id to the label and returns it.
+        """
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_label_name(self, l):
+        """
+        Returns a label's string name given its internal id.
+        If the id does not have a corresponding name, an exception is raised.
+        """
+        for label_name, label_id in self._label_to_id.iteritems():
+            if label_id == l:
+                return label_name
+
+        raise Exception('Label id %d does not exist.' % l)
+
+    def _get_doc_with_label_vector(self, d):
+        """
+        Given a doc id, return the doc and its label bit vector.
+        """
+        return self._w_dp[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """
+        Given a list of labels, return the label bit vector.
+        An empty label list is an alias for all labels being set. This is
```

Comment:
I think this would unnecessarily complicate the two callers as their implementations would no longer be one line (reasonably). I think a compromise would be to still separate the logic, but push logic in instead of pulling it out.